File: sdk/dart/tool/conformance.dart
// Licensed under the Apache License, Version 2.0.
import 'dart:convert';
import 'dart:io';
import 'package:qaraa_client/qaraa_client.dart';

Object? decodeFixture(String schema, Map<String, Object?> raw) {
  switch (schema) {
    case 'corpus':
      return decodeCorpus(raw).toJson();
    case 'observation':
      return RecitationObservation.fromJson(raw).toJson();
    case 'snapshot':
      return ReadingSnapshot.fromJson(raw).toJson();
    case 'command':
      return decodeCommand(raw).toJson();
    case 'event':
    case 'error':
      try {
        final event = decodeEvent(utf8.encode(jsonEncode(raw)));
        return switch (event) {
          SessionCreatedEvent value => value.toJson(),
          SnapshotUpdatedEvent value => value.toJson(),
          SessionDeletedEvent value => value.toJson(),
        };
      } on QaraaException {
        if (raw['type'] == 'error') return raw;
        rethrow;
      }
    default:
      throw FormatException('unknown schema $schema');
  }
}

void main(List<String> args) {
  if (args.length != 2) {
    throw ArgumentError('usage: conformance CONFORMANCE_V1 OUTPUT');
  }
  final root = Directory(args[0]);
  final manifest =
      jsonDecode(File('${root.path}/manifest.json').readAsStringSync())
          as List<Object?>;
  final cases = <Map<String, Object?>>[];
  for (final itemValue in manifest) {
    final item = Map<String, Object?>.from(itemValue! as Map<String, Object?>);
    final raw = Map<String, Object?>.from(
      jsonDecode(File('${root.path}/${item['file']}').readAsStringSync())
          as Map<String, Object?>,
    );
    Object? decoded;
    if (item['valid'] == true) {
      decoded = decodeFixture(item['schema']! as String, raw);
    } else {
      var rejected = false;
      try {
        decodeFixture(item['schema']! as String, raw);
      } on Object {
        rejected = true;
      }
      if (!rejected) {
        throw StateError('invalid fixture accepted: ${item['file']}');
      }
      decoded = null;
    }
    cases.add({
      'fixture': item['file'],
      'decoded': decoded,
      'roundTrip': decoded,
      'errorCode': item['errorCode'],
    });
  }
  File(args[1]).writeAsStringSync(
    const JsonEncoder.withIndent('  ').convert({
      'language': 'dart',
      'sdkVersion': '0.1.0',
      'protocolVersion': 1,
      'cases': cases,
    }),
  );
}

File: sdk/python/src/qaraa/conformance.py
"""Emit normalized fixture results for the cross-language comparator."""
# Licensed under the Apache License, Version 2.0.
from __future__ import annotations
import json, sys
from pathlib import Path
from . import __version__
from .codec import decode_message

def main() -> None:
    if len(sys.argv)!=3: raise SystemExit("usage: python -m qaraa.conformance CONFORMANCE_V1 OUTPUT")
    root,output=Path(sys.argv[1]),Path(sys.argv[2]);manifest=json.loads((root/'manifest.json').read_text())
    cases=[]
    for entry in manifest:
        raw=json.loads((root/entry['file']).read_text())
        if entry['valid']:
            decoded=decode_message(raw,entry['schema']).model_dump(mode='json',by_alias=True,exclude_unset=True)
        else:
            try: decode_message(raw,entry['schema'])
            except Exception: decoded=None
            else: raise RuntimeError(f"invalid fixture accepted: {entry['file']}")
        cases.append({'fixture':entry['file'],'decoded':decoded,'roundTrip':decoded,'errorCode':entry.get('errorCode')})
    output.write_text(json.dumps({'language':'python','sdkVersion':__version__,'protocolVersion':1,'cases':cases},ensure_ascii=False,indent=2))
if __name__=='__main__':main()
